Test that `removePools` does not remove pool metadata.

jonathanknowles opened the pull request:
# Issue Number

#2154 

# Overview

This PR adds documentation and tests that `removePools` does not remove metadata.
    
Strictly speaking, when removing pools, we should also be able to remove any pool metadata that becomes unreachable as a result.

However, this is tricky to get right, since it's possible for more than one pool to share the same metadata hash.

For example, if two pools **_p_** and **_q_** issue registration certificates that share the same metadata hash **_h_**, and then we garbage collect pool **_p_**, we have to be careful to not remove the metadata corresponding to h in order to preserve the metadata for pool **_q_**.

One possible solution would be to remove metadata entries only when we can show that they are no longer reachable: when metadata entries are no longer referenced by any pool.
    
However, for the moment, we adopt the principle of never removing metadata. We can revise this decision in future if the metadata table grows too large.

KtorZ commented on lib/core/src/Cardano/Pool/DB/Sqlite.hs:
```diff
@@ -483,6 +485,15 @@ newDBLayer trace fp timeInterpreter = do
+                mapMaybe (poolRegistrationMetadataHash . entityVal) <$>
+                selectList [ PoolRegistrationPoolId ==. pool ] []
+            forM_ metadataHashesToRemove $ \metadataHash ->
+                deleteWhere [ PoolMetadataHash ==. metadataHash ]
```
Hmmm. That is actually not okay ^^"
Multiple pools could have the same metadata and same metadata hash. This would open up attacks such that pool B registers after pool A with the same metadata, and immediately de-register; clearing up pool's A metadata and making it disappear from the listing.

I suggest instead to add a proper comment to that `removePools` function explaining why metadata are not cleaned up and treated as an immutable data-store.

jonathanknowles replied:
@KtorZ wrote:
> Multiple pools could have the same metadata and same metadata hash.
> This would open up attacks such that pool B registers after pool A with the same metadata, and immediately de-register; clearing up pool's A metadata and making it disappear from the listing.

That's a really good point.

> I suggest instead to add a proper comment to that `removePools` function explaining why metadata are not cleaned up and treated as an immutable data-store.

How about only removing metadata when it is _no longer reachable_?

When removing pool **_p_** with metadata hash **_h_**:

1. First remove all rows corresponding to pool **_p_** from `pool_registrations` and `pool_retirements`, as before.
2. Now find the count **_c_** of all rows in `pool_registrations` with `metadata_hash` = **_h_**.
3. If **_c_** **= 0**, then remove all entries from `pool_metadata` with `metadata_hash` == **_h_**, otherwise leave this table _untouched_.

jonathanknowles replied:
After some discussion, it was decided that we will **not** fix this (for the moment).

Strictly speaking, when removing pools, we should also be able to remove any pool metadata that becomes unreachable as a result (as described in the [above comment](https://github.com/input-output-hk/cardano-wallet/pull/2160#discussion_r492565691)).

However, for the moment, to keep things simple, we adopt the principle of never removing metadata. We can revise this decision in future if the metadata table grows too large.

KtorZ reviewed (changes requested):
Cf previous comment, we cannot clear metadata.

jonathanknowles committed:
Test that `removePools` does not remove metadata.

Strictly speaking, when removing pools, we should also be able to remove any
pool metadata that becomes unreachable as a result.

However, this is tricky to get right, since it's possible for more than one
pool to share the same metadata hash.

For example, if two pools p and q issue registration certificates that share
the same metadata hash h, and then we garbage collect pool p, we have to be
careful to not remove the metadata corresponding to h in order to preserve
the metadata for pool q.

One possible solution would be to remove metadata entries only when we can
show that they are no longer reachable: when metadata entries are no longer
referenced by any pool.

However, for the moment, we adopt the principle of never removing metadata.
We can revise this decision in future if the metadata table grows too large.

jonathanknowles committed:
Add coverage checks to `prop_removePools`.

KtorZ reviewed (approved):
LGTM
